### PR TITLE
Move over to new Azure SDK

### PR DESCRIPTION
**Pull request by tzhanl**
**Purpose**
---------------------------------------------------------------------------------------------------------------
Move over this sample to new Azure SDK:

1. Move the original file to storage-node-v10-ha-ra-grs-v3 folder
2. Add folder storage-node-v10-ha-ra-grs-v4 for the sample referenced to the new SDK
3. Add two sections to the readme:
   - Prerequisites which tell user what info we need
   - Azure SDK versions

**What old sample do**
---------------------------------------------------------------------------------------------------------------
image

**What can be updated to V4:**
---------------------------------------------------------------------------------------------------------------
image

### CODE REVIEW

File: storage-node-v10-ha-ra-grs-v4/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Build a highly available app in Node.js with RA-GRS storage
```
Comment (lilyjma) [inline storage-node-v10-ha-ra-grs-v4/README.md]: I'd say 'a high availability app'

File: storage-node-v10-ha-ra-grs-v4/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS) to create a highly available application that accesses files from secondary storage when there is a problem with primary storage, and then switches back when primary storage becomes available again. For more information, see [Designing HA Apps with RA-GRS storage](https://docs.microsoft.com/azure/storage/common/storage-designing-ha-apps-with-ragrs).
```
Comment (lilyjma) [inline storage-node-v10-ha-ra-grs-v4/README.md]: I'd say, "This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS) to create a high availability app that accesses files from secondary storage when primary storage is down, and then switches back to primary storage when it becomes available again."

File: storage-node-v10-ha-ra-grs-v4/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+1. Add your storage account credentials to the `.env.example` file and then rename it to `.env`.
+
+    ```
+    AZURE_STORAGE_ACCOUNT_NAME=<replace with your storage account name>
+    AZURE_STORAGE_ACCOUNT_ACCESS_KEY=<replace with your storage account access key>
+    ```
+
+    You can find this information in the Azure portal by navigating to your storage account and selecting **Access keys** in the **Settings** section. 
+
+2. Install the required dependencies in the console window by entering `npm install`.
```
Comment (lilyjma) [inline storage-node-v10-ha-ra-grs-v4/README.md]: I'd say 'by running' instead of 'by entering' in Lines 24 and 26

File: README.md [hunk 4]
```diff
@@ -1,93 +1,24 @@
...
-3. Launch the sample from the console window by entering `node index.js`.
-
-4. Wait until the sample reports that the container and file have replicated to secondary storage.
-
-5. When prompted, press D to download the sample file and verify that it comes from primary storage.
-
-To test secondary storage access:
-
-1. Open a command prompt with administrator privileges.
+* Create a storage account.
```
Comment (jongio) [inline README.md:68]: This section needs a header like: In this sample you will do the following:

File: README.md [hunk 5]
```diff
@@ -1,93 +1,24 @@
...
-    ```bash
-    route del <destination_ip> gw <gateway_ip>
-    ```
-
-    **Windows**
-
-    ```cmd
-    route delete <destination_ip>
-    ```
+If you don't have a Microsoft Azure subscription, you can get a FREE trial account <a href="http://go.microsoft.com/fwlink/?LinkId=330212">here</a>.
```
Comment (jongio) [inline README.md]: free, not FREE

File: README.md [hunk 6]
```diff
@@ -1,93 +1,24 @@
...
-10. Press Q to quit the sample and delete the container and file from storage.
+Two folders are referred to different version of Azure SDK.
+* storage-node-v10-ha-ra-grs-v3 referenced to following packages:
+  * [@azure/storage-blob](https://www.npmjs.com/package/@azure/storage-blob/v/10.3.0),10.3.0
+* storage-node-v10-ha-ra-grs-v4 referenced to following packages:
+  * [@azure/storage-blob](https://www.npmjs.com/package/@azure/storage-blob/v/12.0.0-preview.2),12.0.0-preview.2
 
 ## More information
 
 - [About Azure storage accounts](https://docs.microsoft.com/azure/storage/storage-create-storage-account)
```
Comment (jongio) [inline README.md]: Case: Storage Accounts

File: README.md [hunk 7]
```diff
@@ -1,93 +1,24 @@
...
 - [Designing HA Apps with RA-GRS storage](https://docs.microsoft.com/azure/storage/common/storage-designing-ha-apps-with-ragrs)
```
Comment (jongio) [inline README.md]: Case: Storage

File: README.md [hunk 8]
```diff
@@ -1,93 +1,24 @@
...
-    ```cmd
-    route delete <destination_ip>
-    ```
+If you don't have a Microsoft Azure subscription, you can get a FREE trial account <a href="http://go.microsoft.com/fwlink/?LinkId=330212">here</a>.
 
-9. In the console window with the running sample, press D to download the sample file and verify that it comes from primary storage. 
+# Folders introduction
 
-10. Press Q to quit the sample and delete the container and file from storage.
+Two folders are referred to different version of Azure SDK.
```
Comment (jongio) [inline README.md]: Please turn this into a sentence like: You will find the following folders: key-vault-node-quickstart-3, which references the X.X SDK and keyvault-node-quickstart-v4, which uses the X.X version of the SDK.

Something like that.

File: storage-node-v10-ha-ra-grs-v3/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+- azure
+- azure-storage
+description: "This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS)."
+---
+
+# Build a highly available app in Node.js with RA-GRS storage
+
+This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS) to create a highly available application that accesses files from secondary storage when there is a problem with primary storage, and then switches back when primary storage becomes available again. For more information, see [Designing HA Apps with RA-GRS storage](https://docs.microsoft.com/azure/storage/common/storage-designing-ha-apps-with-ragrs).
+
+If you don't have a Microsoft Azure subscription, you can get a FREE trial account <a href="http://go.microsoft.com/fwlink/?LinkId=330212">here</a>.
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v3/README.md]: free, not FREE

File: storage-node-v10-ha-ra-grs-v3/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+## How it works
+
+This sample creates a new container in blob storage and uploads a sample file into it called `HelloWorld.jpg`. It then checks secondary storage repeatedly until both the container and file have been replicated. The user can then enter D to download the file or Q to quit. If the file is successfully downloaded, the sample indicates whether it came from primary or secondary storage. When the user quits the sample, it deletes the container and file. 
+
+## How to run the sample
+
+To run this sample and verify that files are downloaded from primary storage:
+
+1. Add your storage account credentials to the `.env.example` file and then rename it to `.env`.
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v3/README.md]: Please list any pre-reqs at the top of the file.

File: storage-node-v10-ha-ra-grs-v3/README.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    ```
+    AZURE_STORAGE_ACCOUNT_NAME=<replace with your storage account name>
+    AZURE_STORAGE_ACCOUNT_ACCESS_KEY=<replace with your storage account access key>
+    ```
+
+    You can find this information in the Azure portal by navigating to your storage account and selecting **Access keys** in the **Settings** section.
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v3/README.md]: I would put this info above.  You are asking the user for info before you tell them how to get it.

File: storage-node-v10-ha-ra-grs-v3/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+2. Install the required dependencies in the console window by entering `npm install`.
+
+3. Launch the sample from the console window by entering `node index.js`.
+
+4. Wait until the sample reports that the container and file have replicated to secondary storage.
+
+5. When prompted, press D to download the sample file and verify that it comes from primary storage.
+
+To test secondary storage access:
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v3/README.md]: Tell the user what we are going to do before you ask them to execute the steps.  

To test secondary storage access, we are going to...

File: storage-node-v10-ha-ra-grs-v3/README.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    route delete <destination_ip>
+    ```
+
+9. In the console window with the running sample, press D to download the sample file and verify that it comes from primary storage. 
+
+10. Press Q to quit the sample and delete the container and file from storage.
+
+## More information
+
+- [About Azure storage accounts](https://docs.microsoft.com/azure/storage/storage-create-storage-account)
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v3/README.md]: Case: Storage Accounts

File: storage-node-v10-ha-ra-grs-v4/README.md (hunk 1 above)
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v4/README.md]: All of my comments on v3.md apply to v4.md.

Comment (jongio) [inline storage-node-v10-ha-ra-grs-v4/README.md]: Please merge v3.md and v4.md into a single file and notate any differences between the SDKs, which should be minimal.

File: storage-node-v10-ha-ra-grs-v4/index.js [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  input: process.stdin,
+  output: process.stdout
+});
+
+// Read environment variables for storage access.
+const accountName = process.env.AZURE_STORAGE_ACCOUNT_NAME;
+const storageAccessKey = process.env.AZURE_STORAGE_ACCOUNT_ACCESS_KEY;
+const sharedKeyCredential = new SharedKeyCredential(accountName, storageAccessKey);
+const primaryAccountURL = `https://${accountName}.blob.core.windows.net`;
+const secondaryAccountURL = `https://${accountName}-secondary.blob.core.windows.net`;
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v4/index.js]: Is -secondary a standard naming convention or is this something that the sample requires.  If the user has to do anything to set this up, then you need to let them know that they need to name it this way.

File: storage-node-v10-ha-ra-grs-v4/index.js [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+  console.log("Ready for blob download. Enter (D) to download or (Q) to quit, followed by ENTER.");
+  readInput.on('line', async (input) => {
+    switch (input.toUpperCase()) {
+      case 'D':
+        console.log("Attempting to download blob...");
+        await downloadBlob();
+        break;
+      case 'Q':
+        console.log("Exiting...");
+        readInput.close();
```
Comment (jongio) [inline storage-node-v10-ha-ra-grs-v4/index.js]: Why close here and not the in D case?

**Review by jongio — CHANGES_REQUESTED** [review]

File: storage-node-v10-ha-ra-grs-v4/index.js (hunk 15 above)
Comment (tzhanl) [reply storage-node-v10-ha-ra-grs-v4/index.js]: readInput is to read user input "D" or "Q" here. It is not needed after people input "Q", that's why it is better to close readInput in Q case, not the D case.

**Review by jongio — APPROVED** [review]

File: README.md [hunk 16]
```diff
@@ -1,27 +1,14 @@
...
-- nodejs
-products:
-- azure
-- azure-storage
-description: "This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS)."
----
+# Build a high availability app in Node.js with RA-GRS storage
 
-# Build a highly available app in Node.js with RA-GRS storage
+This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS) to create a high availability app that accesses files from secondary storage when primary storage is down, and then switches back to primary storage when it becomes available again. For more information, see [Designing HA Apps with RA-GRS storage](https://docs.microsoft.com/azure/storage/common/storage-designing-ha-apps-with-ragrs).
```
Comment (jongio) [inline README.md]: highly available, not high availability

File: README.md [hunk 17]
```diff
@@ -1,27 +1,14 @@
...
 
-This sample shows how to use the Node.js V10 Storage SDK with read-access geo-redundant storage (RA-GRS) to create a highly available application that accesses files from secondary storage when there is a problem with primary storage, and then switches back when primary storage becomes available again. For more information, see [Designing HA Apps with RA-GRS storage](https://docs.microsoft.com/azure/storage/common/storage-designing-ha-apps-with-ragrs).
+## SDK Versions
 
-If you don't have a Microsoft Azure subscription, you can get a FREE trial account <a href="http://go.microsoft.com/fwlink/?LinkId=330212">here</a>.
+You will find the following folders: [storage-node-ha-ra-grs-v10](./storage-node-ha-ra-grs-v10), which references the [@azure/storage-blob version 10.3.0](https://www.npmjs.com/package/@azure/storage-blob/v/10.3.0) of the SDK and [storage-node-ha-ra-grs-v12](./storage-node-ha-ra-grs-v12), which uses the [@azure/storage-blob version 12.0.0](https://www.npmjs.com/package/@azure/storage-blob/v/12.0.0) of the SDK.
```
Comment (jongio) [inline README.md]: See my format comment on key-vault and use similar format here.

**Review by jongio — CHANGES_REQUESTED** [review]

File: README.md [hunk 18]
```diff
@@ -11,36 +11,33 @@ description: "This sample shows how to use the Node.js V10 Storage SDK with read
...
+## Prerequisites
+
+Following prerequisites are needed to run this sample and verify that files are downloaded from primary storage:
 
 1. Add your storage account credentials to the `.env.example` file and then rename it to `.env`.
 
     ```
-    AZURE_STORAGE_ACCOUNT_NAME=<replace with your storage account name>
-    AZURE_STORAGE_ACCOUNT_ACCESS_KEY=<replace with your storage account access key>
+    AZURE_STORAGE_ACCOUNT_NAME=<Replace With Your Storage Account Name>
```
Comment (jongio) [inline README.md]: In the last sample you used 'screaming caps' in the replacement token and here you use sentences.  Please be consistent across samples.

**Review by jongio — APPROVED** [review]
Address the one last comment and then approved.

**Review by jongio — APPROVED** [review]

**Pull request merged**